### PR TITLE
SIMSBIOHUB-471: Updates to Critterbase to support ITIS

### PR DESCRIPTION
**Pull request by curtisupshall**
## Links to Jira Tickets

- [SIMSBIOHUB-471](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-471)

## Description of Changes

- Mostly refactored existing code around the 'Manage Animals' page.
- Critterbase changed from providing a list of many 'detailed critters' to only a single 'detailed critter'
### Net-new Components
- AnimalSection - Acts as a switch for the individual animal sections. ie: 'Markings' || 'Captures' || Mortality ...
- AnimalSectionWrapper - Wrapper component for the styling around the body component of the 'Manage Animals' page

## Testing Notes

- Should be able to add a critter + all related attribute sections

### CODE REVIEW

File: database/src/seeds/03_basic_project_survey_setup.ts [hunk 1]
```diff
@@ -5,8 +5,8 @@ const DB_SCHEMA = process.env.DB_SCHEMA;
 const DB_SCHEMA_DAPI_V1 = process.env.DB_SCHEMA_DAPI_V1;
 const PROJECT_SEEDER_USER_IDENTIFIER = process.env.PROJECT_SEEDER_USER_IDENTIFIER;
 
-const NUM_SEED_PROJECTS = Number(process.env.NUM_SEED_PROJECTS ?? 2);
-const NUM_SEED_SURVEYS_PER_PROJECT = Number(process.env.NUM_SEED_SURVEYS_PER_PROJECT ?? 2);
+const NUM_SEED_PROJECTS = Number(process.env.NUM_SEED_PROJECTS ?? 1);
```
Comment (curtisupshall) [inline database/src/seeds/03_basic_project_survey_setup.ts]: The  `env.docker`  says that these two values default to 2. So we should either revert this change, or amend that comment.

File: api/src/openapi/schemas/critter.ts [hunk 2]
```diff
@@ -9,16 +9,30 @@ const critterSchema: OpenAPIV3.SchemaObject = {
...
+      type: 'string',
+      nullable: true
     },
     wlh_id: {
-      type: 'string'
+      type: 'string',
+      nullable: true
     },
     itis_tsn: {
+      type: 'number'
```
Comment (curtisupshall) [inline api/src/openapi/schemas/critter.ts]: Nit: `type: 'integer'`

File: api/src/openapi/schemas/critter.ts [hunk 3]
```diff
@@ -190,6 +204,30 @@ const quantitativeMeasurmentSchema: OpenAPIV3.SchemaObject = {
...
+      format: 'uuid'
+    },
+    animal_id: {
+      type: 'string'
+    },
+    wlh_id: {
+      type: 'string'
+    },
+    itis_tsn: {
+      type: 'number'
```
Comment (curtisupshall) [inline api/src/openapi/schemas/critter.ts]: Nit: `type: 'integer'`

File: app/src/features/surveys/telemetry/ManualTelemetryList.tsx [hunk 4]
```diff
@@ -151,6 +146,8 @@ const ManualTelemetryList = () => {
   };
 
   const handleSubmit = async (data: AnimalDeployment) => {
+    console.log(data);
```
Comment (curtisupshall) [inline app/src/features/surveys/telemetry/ManualTelemetryList.tsx]: Remove

File: app/src/features/surveys/view/survey-animals/AnimalList.tsx [hunk 5]
```diff
@@ -123,10 +149,10 @@ const AnimalList = (props: IAnimalListProps) => {
           sx={{
             background: grey[100]
           }}>
-          {!sortedCritterData.length ? (
-            <ListPlaceholder displaySkeleton={!!isLoading && !sortedCritterData?.length} />
+          {!surveyCritters?.length ? (
+            <ListPlaceholder displaySkeleton={!!isLoading && !surveyCritters?.length} />
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/AnimalList.tsx]: Don't need the double negation here (`!!`)

File: app/src/features/surveys/view/survey-animals/AnimalList.tsx [hunk 6]
```diff
@@ -138,11 +164,11 @@ const AnimalList = (props: IAnimalListProps) => {
                   }
                 }}
                 key={critter.critter_id}
-                expanded={critter.survey_critter_id.toString() === survey_critter_id}>
+                expanded={critter.survey_critter_id === Number(survey_critter_id)}>
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/AnimalList.tsx]: I would move the cast to number during assignment:

```javascript
const { cid } = useQuery();
const survey_critter_id = Number(cid);
```

File: app/src/features/surveys/view/survey-animals/SurveyAnimalsPage.tsx [hunk 7]
```diff
@@ -1,324 +1,82 @@
...
 export const SurveyAnimalsPage = () => {
-  const [selectedSection, setSelectedSection] = useState<IAnimalSections>(SurveyAnimalsI18N.animalGeneralTitle);
-  const { cid: survey_critter_id } = useQuery();
-  const [openAddDialog, setOpenAddDialog] = useState(false);
   const bhApi = useBiohubApi();
-  const telemetryApi = useTelemetryApi();
-  const dialogContext = useContext(DialogContext);
-  const { surveyId, projectId, artifactDataLoader } = useContext(SurveyContext);
+  const cbApi = useCritterbaseApi();
+  const { cid: survey_critter_id } = useQuery();
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/SurveyAnimalsPage.tsx]: Same comment here re: casting to Number

File: app/src/features/surveys/view/survey-animals/form-sections/FamilyAnimalForm.tsx [hunk 8]
```diff
@@ -38,143 +41,213 @@ const useStyles = makeStyles((theme: Theme) => ({
...
+                      </li>
+                    ))}
+                  </ul>
+                </Paper>
+                <Paper variant="outlined" sx={{ p: 1, mb: 2 }}>
+                  <Typography component="h4">Children:</Typography>
+                  <ul>
+                    {familyHierarchy?.children.map(
+                      (
+                        a: { critter_id: string; animal_id: string } //I will type this better I promise
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/form-sections/FamilyAnimalForm.tsx]: 👀

File: app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx [hunk 9]
```diff
@@ -74,81 +104,117 @@ export const MeasurementAnimalFormContent = (props: MeasurementFormContentProps)
...
+                name={'measurement_comment'}
+              />
+            </Grid>
+          </Grid>
+        )
+      }}
+    />
   );
 };
+// <CbSelectField
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx]: Remove

File: app/src/hooks/api/useSurveyApi.ts [hunk 10]
```diff
@@ -494,6 +429,15 @@ const useSurveyApi = (axios: AxiosInstance) => {
...
     }
     const flattened = { ...body, ...body.deployments[0] };
+
+    delete flattened.deployment_id;
+    delete flattened.deployments;
+    if (!flattened.device_model) {
+      delete flattened.device_model;
+    }
+
+    console.log(flattened);
```
Comment (curtisupshall) [inline app/src/hooks/api/useSurveyApi.ts]: Remove

**Review by al-rosenthal — COMMENTED** [review]
I tested this locally and things work well. I just noticed a couple of small things:
Add Critter Dialog
- The taxon search has an arrow icon and a help icon overlapping
- The dialog changes width when sex is selected

Add Marking
- Might just be a data issue but I'm unable to complete this form. No body locations are available for moose, caribou or wolf and the form cannot be completed without a body location

Family
- Might be intended but when adding a critter to a family the family page doesn't update until both child and parent are present

**Comment by NickPhura** [comment]
Issues with Family records.

Steps:

1. Create new animal
2. Add family
3. Select existing family (A1111-Moose_family)
4. Select 'parent'
5. Save

- POST http://localhost:6107/api/critterbase/family/parents
- Body: `{"family_id":"059a6f0e-9a39-488b-b473-97976776abd7","family_label":"","parent_critter_id":"bab99e32-a19f-47ab-a2aa-cbd2a63f787e"}`
- Response: `{"family_id":"059a6f0e-9a39-488b-b473-97976776abd7","parent_critter_id":"bab99e32-a19f-47ab-a2aa-cbd2a63f787e","create_user":"12bd1aa7-c3fa-4e5c-bca6-1b4ace72eb89","update_user":"12bd1aa7-c3fa-4e5c-bca6-1b4ace72eb89","create_timestamp":"2024-03-26T17:59:30.373Z","update_timestamp":"2024-03-26T17:59:30.373Z"}`

Outcome:

- Record is shown in the UI with the correct family name, but the incorrect relationship (shows 'Child')
image

---

Attempting to Edit the record:

- Family field is just a text field (no dropdown options)
- Relationship field is greyed out
image

- I can still edit the name of the family, and a successful PATCH request is sent

PATCH: http://localhost:6107/api/critterbase/family/059a6f0e-9a39-488b-b473-97976776abd7
Body: `{"family_label":"A1111-Moose_family 222"}`
Response: `{"family_id":"059a6f0e-9a39-488b-b473-97976776abd7","family_label":"A1111-Moose_family 222","create_user":"12bd1aa7-c3fa-4e5c-bca6-1b4ace72eb89","update_user":"12bd1aa7-c3fa-4e5c-bca6-1b4ace72eb89","create_timestamp":"2023-10-13T16:51:55.888Z","update_timestamp":"2023-10-13T16:51:55.888Z"}`

image

---

Attempting to Delete the record fails with a 404:

- DELETE http://localhost:6107/api/critterbase/family/children
- Body: `{"family_id":"059a6f0e-9a39-488b-b473-97976776abd7","child_critter_id":"bab99e32-a19f-47ab-a2aa-cbd2a63f787e"}`
- Response: `{"error":"Record to delete does not exist.","issues":[{"cause":"Record to delete does not exist."}]}`

**Comment by MacDeluca** [comment]
> I tested this locally and things work well. I just noticed a couple of small things: Add Critter Dialog
> 
> * The taxon search has an arrow icon and a help icon overlapping

Im gonna leave this temporarily. There's quite a few components that are using this tooltip. With it using abosolute positioning, its probably more work than its worth to modify. Tech debt.

> * The dialog changes width when sex is selected

This is fixed

> Add Marking
> 
> * Might just be a data issue but I'm unable to complete this form. No body locations are available for moose, caribou or wolf and the form cannot be completed without a body location

I think this is a data issue, I have recently updated live DEV to apply most attributes to the Animalia Kingdown, so all critters will inherit.

> Family
> 
> * Might be intended but when adding a critter to a family the family page doesn't update until both child and parent are present

Resolved a bunch of issues with this section, this included.

**Review by al-rosenthal — DISMISSED** [review]
🫀

**Comment by al-rosenthal** [comment]
Noticed another small thing, the comments in the markings section don't appear to be getting saved. It looks like it is missing the in the update payload

**Review by al-rosenthal — APPROVED** [review]
🐶

**Review by MacDeluca — APPROVED** [review]
Approving on behalf of @curtisupshall
